src/Type/Generic/GenericObjectType.php

```diff
@@ -135,22 +135,25 @@ public function inferTemplateTypes(Type $receivedType): TemplateTypeMap
 			return $receivedType->inferTemplateTypesOn($this);
 		}
 
-		if (
-			$receivedType instanceof GenericObjectType
-			&& $receivedType->getClassName() === $this->getClassName()
-		) {
-			$otherTypes = $receivedType->getTypes();
-			$typeMap = TemplateTypeMap::createEmpty();
-
-			foreach ($this->getTypes() as $i => $type) {
-				$other = $otherTypes[$i] ?? new ErrorType();
-				$typeMap = $typeMap->union($type->inferTemplateTypes($other));
-			}
+		if (!$receivedType instanceof ObjectType) {
```

Comment:
Maybe TypeWithClassName would be better, to accomodate StaticType? It has the same getClassName() method. (There might be other instances of ObjectType in generic code that would be better to replace with StaticType.)